File: solana-lens/scripts/collectors/helius_client.py
import requests
import json
from typing import Dict, List, Any, Optional, Union

class HeliusClient:
    """
    Client for interacting with the Helius API for Solana blockchain data.
    
    Provides methods for querying account information, transactions, 
    and other Solana blockchain data.
    """
    
    def __init__(self, api_key: str):
        """
        Initialize the Helius API client.
        
        Args:
            api_key (str): API key for authentication
        """
        self.api_key = api_key
        self.base_url = f"https://mainnet.helius-rpc.com/?api-key={self.api_key}"
        self.headers = {
            "Content-Type": "application/json"
        }
    
    def _make_request(self, method: str, params: List[Any]) -> Dict:
        """
        Make a JSON-RPC request to the Helius API.
        
        Args:
            method (str): The RPC method to call
            params (List[Any]): Parameters for the RPC method
            
        Returns:
            Dict: Response from the API
            
        Raises:
            Exception: If the API request fails
        """
        payload = {
            "jsonrpc": "2.0",
            "id": 1,
            "method": method,
            "params": params
        }
        
        try:
            response = requests.post(self.base_url, headers=self.headers, json=payload)
            response.raise_for_status()
            result = response.json()
            
            if "error" in result:
                raise Exception(f"API error: {json.dumps(result['error'])}")
                
            return result
        except requests.exceptions.RequestException as e:
            error_msg = f"API request failed: {str(e)}"
            if hasattr(e, 'response') and e.response is not None:
                try:
                    error_data = e.response.json()
                    error_msg += f". Details: {json.dumps(error_data)}"
                except:
                    error_msg += f". Status code: {e.response.status_code}"
            raise Exception(error_msg)
    
    # Account & Balance Endpoints
    
    def get_account_info(self, address: str, encoding: str = "jsonParsed") -> Dict:
        """
        Get information for a specific account.
        
        Args:
            address (str): The account address
            encoding (str, optional): Response encoding. Defaults to "jsonParsed".
            
        Returns:
            Dict: Account information
        """
        params = [address, {"encoding": encoding}]
        return self._make_request("getAccountInfo", params)
    
    def get_balance(self, address: str) -> Dict:
        """
        Get the SOL balance of an account.
        
        Args:
            address (str): The account address
            
        Returns:
            Dict: Account balance information
        """
        params = [address]
        return self._make_request("getBalance", params)
    
    def get_token_account_balance(self, token_account: str) -> Dict:
        """
        Get the token balance for a specific token account.
        
        Args:
            token_account (str): The token account address
            
        Returns:
            Dict: Token balance information
        """
        params = [token_account]
        return self._make_request("getTokenAccountBalance", params)
    
    def get_token_accounts_by_owner(self, owner: str, program_id: str = "TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA") -> Dict:
        """
        Get all token accounts for an owner.
        
        Args:
            owner (str): The owner's address
            program_id (str, optional): Token program ID. Defaults to SPL Token program.
            
        Returns:
            Dict: List of token accounts
        """
        params = [
            owner, 
            {"programId": program_id}, 
            {"encoding": "jsonParsed"}
        ]
        return self._make_request("getTokenAccountsByOwner", params)
    
    # Transaction Endpoints
    
    def get_transaction(self, signature: str, encoding: str = "jsonParsed") -> Dict:
        """
        Get transaction details.
        
        Args:
            signature (str): Transaction signature
            encoding (str, optional): Response encoding. Defaults to "jsonParsed".
            
        Returns:
            Dict: Transaction details
        """
        params = [
            signature, 
            {"encoding": encoding, "maxSupportedTransactionVersion": 0}
        ]
        return self._make_request("getTransaction", params)
    
    def get_signatures_for_address(self, address: str, limit: int = 100) -> Dict:
        """
        Get signatures for transactions involving an address.
        
        Args:
            address (str): The account address
            limit (int, optional): Maximum number of signatures. Defaults to 100.
            
        Returns:
            Dict: List of transaction signatures
        """
        params = [address, {"limit": limit}]
        return self._make_request("getSignaturesForAddress", params)
    
    def simulate_transaction(self, serialized_tx: str) -> Dict:
        """
        Simulate executing a transaction.
        
        Args:
            serialized_tx (str): Serialized transaction
            
        Returns:
            Dict: Simulation results
        """
        params = [
            serialized_tx, 
            {"encoding": "base64", "commitment": "processed"}
        ]
        return self._make_request("simulateTransaction", params)
    
    # Program Endpoints
    
    def get_program_accounts(self, program_id: str, filters: List[Dict] = None) -> Dict:
        """
        Get all accounts owned by a program.
        
        Args:
            program_id (str): Program ID
            filters (List[Dict], optional): Filters to apply. Defaults to None.
            
        Returns:
            Dict: List of program accounts
        """
        params = [
            program_id, 
            {"encoding": "jsonParsed", "filters": filters or []}
        ]
        return self._make_request("getProgramAccounts", params)

File: solana-lens/scripts/api/helius-client.js
/**
 * Helius API Client
 * 
 * Client for interacting with Helius API to fetch Solana data
 */

import fetch from 'isomorphic-fetch';

export class HeliusClient {
  /**
   * Create a new HeliusClient
   * @param {string} apiKey - Helius API key
   * @param {string} baseUrl - Base URL for the Helius API
   */
  constructor(apiKey, baseUrl = 'https://api.helius-rpc.com/v0') {
    this.apiKey = apiKey;
    this.baseUrl = baseUrl;
    this.isDemoMode = apiKey === 'demo';
    
    if (this.isDemoMode) {
      console.warn('HeliusClient initialized in demo mode. API functionality will be limited.');
    }
  }

  /**
   * Get transaction details for a given signature
   * @param {string} signature - Transaction signature
   * @returns {Promise<Object>} - Transaction details
   */
  async getTransaction(signature) {
    return this._makeRequest(`/transactions/${signature}`);
  }

  /**
   * Get transaction details for a given signature (async version)
   * @param {string} signature - Transaction signature
   * @returns {Promise<Object>} - Transaction details
   */
  async getTransactionAsync(signature) {
    return this.getTransaction(signature);
  }

  /**
   * Get signatures for an address
   * @param {string} address - Account address
   * @param {Object} options - Options for the request
   * @returns {Promise<Object>} - Signature information
   */
  async getSignaturesForAddress(address, options = {}) {
    const params = new URLSearchParams(options).toString();
    return this._makeRequest(`/addresses/${address}/transactions?${params}`);
  }

  /**
   * Get signatures for an address (async version)
   * @param {string} address - Account address
   * @param {Object} options - Options for the request
   * @returns {Promise<Object>} - Signature information
   */
  async getSignaturesForAddressAsync(address, options = {}) {
    return this.getSignaturesForAddress(address, options);
  }

  /**
   * Get parsed transaction details with enhanced data
   * @param {string} signature - Transaction signature
   * @returns {Promise<Object>} - Enhanced transaction details
   */
  async getEnhancedTransaction(signature) {
    return this._makeRequest(`/transactions/parsed/${signature}`);
  }

  /**
   * Search for transactions by specified filters
   * @param {Object} query - Search query parameters
   * @returns {Promise<Array>} - Matching transactions
   */
  async searchTransactions(query) {
    return this._makeRequest('/transactions/search', 'POST', query);
  }

  /**
   * Make a request to the Helius API
   * @private
   * @param {string} path - API endpoint path
   * @param {string} method - HTTP method (GET, POST, etc)
   * @param {Object} body - Request body for POST requests
   * @returns {Promise<any>} - API response
   */
  async _makeRequest(path, method = 'GET', body = null) {
    if (this.isDemoMode) {
      return this._getMockResponse(path, method, body);
    }

    const url = `${this.baseUrl}${path}${path.includes('?') ? '&' : '?'}api-key=${this.apiKey}`;
    const options = {
      method,
      headers: {
        'Content-Type': 'application/json'
      }
    };

    if (body) {
      options.body = JSON.stringify(body);
    }

    try {
      const response = await fetch(url, options);
      
      if (!response.ok) {
        throw new Error(`Helius API responded with status: ${response.status}`);
      }
      
      return await response.json();
    } catch (error) {
      console.error(`Helius API request failed: ${error.message}`);
      throw error;
    }
  }

  /**
   * Provide mock responses for demo mode
   * @private
   * @param {string} path - API endpoint path
   * @param {string} method - HTTP method
   * @param {Object} body - Request body
   * @returns {Promise<Object>} - Mock response data
   */
  async _getMockResponse(path, method, body) {
    // Return limited mock data for demo mode
    if (path.includes('/transactions/search')) {
      return {
        result: [
          // Sample transaction data
          { /* mock transaction data */ }
        ]
      };
    }
    
    if (path.includes('/transactions/')) {
      return {
        result: { /* mock single transaction data */ }
      };
    }
    
    if (path.includes('/addresses/')) {
      return {
        result: [
          // Sample signatures
          { signature: 'mock_signature_1', blockTime: Date.now() - 100000 },
          { signature: 'mock_signature_2', blockTime: Date.now() - 200000 }
        ]
      };
    }
    
    return { result: [] }; // Default empty response
  }
}
